### ISSUE TITLE
Reference counting issues on macOS?

### ISSUE THREAD

**Issue by lysannschlegel** [GameFoundry/bsf#180]
I'm just getting started using bsf, but I've run into some weird issues on macOS that seem to be related to the resource handles.

I've made a gist with the code that I'm using: https://gist.github.com/lysannschlegel/fff34e321471fdb42633587d31a0c30d

If I run this exact code, the application appears to deadlock like this: https://gist.github.com/lysannschlegel/fff34e321471fdb42633587d31a0c30d#gistcomment-2648953

What's curious about this is that the deadlock is caused when trying to delete a shader that is a dependency for a material for which `bs::Resource::areDependenciesLoaded` is currently being called. The shader in question has `mRefCount = 1` after it was added to the vector used in `areDependenciesLoaded`, and therefore gets deleted when the vector goes out of scope. From what I understand however, the shader should be referenced in addition by:
* the handle in my `runApplication` function
* the reference from the material (`mShader`)
* an internal reference that is made when the resource is loaded (?)

When running the exact same code on my Windows machine it works fine, and I've stepped through the reference count changes a bit (using a data breakpoint) and it all appears to make sense to me.
(I haven't yet tried stepping through it on macOS because the lldb isn't showing things very well :/)

Just a hunch: Looking into the resource handle code I'm just wondering whether it's a good idea to alter the `mRefCount` like that, perhaps the implementations on different platforms might cause issues with that?

I also want to point out that if I change this code slightly depending on the change it sometimes works, and sometimes it causes other weird issues (usually crashes), like a pure virtual function call, nothing being rendered, etc. These aren't 100% reproducible though, and if it's really a memory management issue it's likely that the root cause is the same.

**Comment by BearishSun** [comment]
Thanks for the detailed report. I've seen some weirdness with resource handle reference counting recently but haven't had a chance to track down the problem yet.

I'll take a closer look as soon as I get a chance, but until I do you can try this:

Modify the `TResourceHandleBase<false>::releaseRef` code to this (`BsResourceHandle.h`, line 163):
```
void releaseRef() 
{ 
	if (mData)
	{
		std::uint32_t refCount = mData->mRefCount.fetch_sub(1, std::memory_order_release);

		if (refCount == 1)
		{
			std::atomic_thread_fence(std::memory_order_acquire);
			destroy();
		}
	}
};
```

Let me know if that changes anything.

**Comment by BearishSun** [comment]
Also try changing the move assignment operator to this:
```
TResourceHandle& operator=(TResourceHandle&& other)
{
	if(this == &other)
		return *this;

	this->releaseRef();
	this->mData = std::exchange(other.mData, nullptr);

	return *this;
}
```

That seems like the more likely cause.

**Comment by lysannschlegel** [comment]
Unfortunately still the same issue with both changes. I've tried them both in isolation and together.

**Comment by BearishSun** [comment]
I did end up experiencing the issue myself on Windows. It wasn't quite in the same manner, but it was also failing to properly count references, so it's a good chance this was your issue as well. If you get a chance let me know if that resolves it for you. The fix is in 895bec84fbfcbf204c0ca0f8effb7adefeb679bb.

**Comment by lysannschlegel** [comment]
I just tested again with the latest master and everything seems fine. Thanks for your help!
